grouping optimization for linematch algorithm

jwhite510 opened the pull request:
@lewis6991 
A pull request for the grouping optimization

jwhite510 committed:
feat(diff): grouping optimization for linematch algorithm

lewis6991 commented:
CI is failing. Need to fix the `clint.py` errors and also run `make formatc`.

jwhite510 commented:
I'm questioning if this is worth adding at this point, because doing this grouping optimization requires removing the space optimization, and without that, the memory allocated to run the algorithm gets much larger.
without the optimization:
2 files O(n^2)
3 files O(n^3)
4 files O(n^4)
with the optimization
2 files O(n)
3 files O(n^2)
4 files O(n^3)

lewis6991 commented:
Since this produces much nicer diffs, I would have thought this is worth adding. Out of all your PR's, this is the one I'm most interesting in.

jwhite510 commented on src/nvim/linematch.c:
```diff
@@ -354,34 +335,67 @@ size_t linematch_nbuffers(const char **diff_blk, const int *diff_len, const size
   }
 
   // create the flattened path matrix
   diffcmppath_T *diffcmppath = xmalloc(sizeof(diffcmppath_T) * memsize);
```
this is where it will crash, `memsize` gets much larger. any idea on a trigger to disable the grouping optimization when some limit is exceeded? something like "if `memsize` gets larger than some value, disable the grouping optimization and use the space optimization"

jwhite510 commented:
I changed the function that searches for the optimal path, it's now a complete tree traversal with memoization to prevent retracing the same path. The other solution worked sometimes but I found not for all cases, I've added a test case would have failed previously.

justinmk commented:
what can we say about this in news.txt ?

lewis6991 commented:
It's explained here: https://github.com/neovim/neovim/pull/23569

Basically the scoring algorithm now favours larger and less groups (as I understand).

tmillr commented:
It might be nice to have the option to increase the score (or set a configurable multiplier) for matches earlier in the line (or for a whole-word match that starts the line). I noticed in a diff recently, in a lua table literal where a leading key/field doesn't change (but it's value after `=` is larger and does change/move), gets lined up with the wrong key (even though the key didn't change at all).

Idk if it has anything to do with this pr or not, but for example I recently ran into something like the following:

```
- someKey = someLongLongLongLongLongLongLongValue,  |  + Key = someLongLongLongLongLongLongLongValue,
                                                    |  + someKey = somethingNew,
```

I might prefer this instead:

```
+                                                   |  + Key = someLongLongLongLongLongLongLongValue,
- someKey = someLongLongLongLongLongLongLongValue,  |  + someKey = somethingNew,
```
